gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java

```diff
@@ -187,12 +207,31 @@ protected void startUp() {
   synchronized void offer(Dag<JobExecutionPlan> dag) throws IOException {
     //Persist the dag
     this.dagStateStore.writeCheckpoint(dag);
-    //Add it to the queue of dags
-    if (!this.queue.offer(dag)) {
+    long flowExecutionId = Long.parseLong(dag.getNodes().get(0).getValue().getJobSpec().getConfigAsProperties().getProperty(ConfigurationKeys.FLOW_EXECUTION_ID_KEY));
+    int queueId = (int) (flowExecutionId % this.numThreads);
+    // Add the dag to the specific queue determined by flowExecutionId
+    if (!this.queue[queueId].offer(dag)) {
       throw new IOException("Could not add dag" + DagManagerUtils.generateDagId(dag) + "to queue");
     }
   }
 
+  /**
+   * Method to submit a {@link URI} for canellation requsts to the {@link DagManager}.
+   * The {@link DagManager} adds the dag to the {@link BlockingQueue} to be picked up by one of the {@link DagManagerThread}s.
+   */
+ synchronized public void offer(URI uri) {
+    String flowGroup = FlowConfigResourceLocalHandler.FlowUriUtils.getFlowGroup(uri);
+    String flowName = FlowConfigResourceLocalHandler.FlowUriUtils.getFlowName(uri);
+
+    List<Long> flowExecutionIds = this.jobStatusRetriever.getLatestExecutionIdsForFlow(flowName, flowGroup, 10);
```

Comment:
Can we make the number of executions to cancel configurable with a default of 1?

Reply:
Returning less than the required flowExecutionIds is a bigger problem than returning more than required.
In former case, we won't be able to cancel old stuck flows

Reply:
Fair.